src/ser/mod.rs

```diff
@@ -40,18 +39,10 @@ impl From<u8> for Error {
     }
 }
 
-impl error::Error for Error {
-    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
-        None
-    }
-
-    fn description(&self) -> &str {
-        "(use display)"
-    }
-}
+impl serde::ser::StdError for Error {}
```

Comment:
D’oh, yes, you’re right.  I’ve misread what serde is doing.  Here’s what it
does:

```rust
#[cfg(all(feature = "unstable", not(feature = "std")))]
pub use core::error::Error as StdError;
#[cfg(feature = "std")]
pub use std::error::Error as StdError;
#[cfg(not(any(feature = "std", feature = "unstable")))]
pub use std_error::Error as StdError;
```

If either `std` or `unstable` feature is enabled, `StdError` will implement
`std::error::Error`.  `unstable` requires nightly compiler.

I thought I had it figured out but turns out std::error::Error is still major
PITA.  (To be honest, as far as I’m concerned, std::error::Error is spawn of the
devil which should never be used).

How about adding `unstable` feature to serde-json-wasm.  If enabled, it would
enable error_in_core feature and use core::error::Error.  That would make the
library no_std but crucially would change nothing for people who don’t care
about no_std?

Reply:
So, adding `unstable` would make it `no_std`, but will only work with the nightly compiler, right?

Reply:
Yes.  There is no way to implement `std::error::Error` and not require nightly.  (I thought there was alas no).  This should be a workable solution.  Users who care about `no_std` can enable the feature.  Users who don’t, don’t need to do anything and their code will work even if some other crate enables `unstable`.

A different option is to add `no_std` feature and when it’s enabled `std::error::Error` would not be implemented but that means that compilation of code which is unaware of this feature might end up breaking if it expects that trait to be implemented.

Of course ideally `serde-json-wasm` would have a `std` default feature but as I understand you guys aren’t keen on affecting `default-features = false` builds.  (To be honest I think this is too strong of a position — not even Rust guarantees that upgrades will be completely hassle-free — but it’s not really for me to say).

Honestly, as cosmwasm is gearing for 2.0 release I’d be more than happy with ditching std::error::Error altogether but that’s nighter here nor there.

Reply:
I am not a fan of a `no_std` feature. Cargo features are supposed to be additive.
As I mentioned, we will release `serde-json-wasm 1.0` once this PR is in, so we can do breaking changes here.
Given that, I think an `std` feature (and maybe an additional `unstable` feature) is the cleanest solution.

Reply:
What’s the schedule for 1.0 release?

The main reason I’m trying to make this non-breaking is that I’d love to have this released as a minor or patch release so that I could start using this even if other crates I’m using aren’t migrated to 1.0 yet.  Currently to be no_std I have to maintain forks and patch dependencies.  If there was serde-json-wasm with `unstable` I’d be able to just enable that feature and everything would work.

Though I guess #59 complicates this now anyway?

Reply:
If we can get this PR in today, I can release 1.0 today. Otherwise, it will have to wait a bit more, as I'm on vacation next week.
If you want to make it non-breaking with the unstable feature, then I would create a release branch from before #59, we merge into that and release a patch version from that, so yes it gets a bit more complicated.

Reply:
SG. I’ll prepare a new PR within a couple hours.  I also sent #60 and #61 If you want to approve and merge those this one will be smaller.

Reply:
Done.

```
$ cargo build
    Finished dev [unoptimized] target(s) in 0.00s
$ cargo +nightly build --target thumbv6m-none-eabi --no-default-features
    Finished dev [unoptimized] target(s) in 0.04s
$ cargo +nightly build --target thumbv6m-none-eabi --no-default-features --features unstable
    Finished dev [unoptimized] target(s) in 0.00s
$ cargo test
    Finished test [unoptimized] target(s) in 0.10s
     Running unittests src/lib.rs (target/debug/deps/serde_json_wasm-d19a6832c1bd7bc4)

running 54 tests
test de::tests::array ... ok
test de::tests::bool ... ok
test de::tests::deserialize_optional_vector ... ok
test de::tests::deserialize_embedded_enum ... ok
test de::tests::enum_clike ... ok
test de::tests::ignoring_extra_fields ... ok
test de::tests::integer128 ... ok
test de::tests::newtypes ... ok
test de::tests::parse_whitespace ... ok
test de::tests::string ... ok
test de::tests::struct_bool ... ok
test de::tests::struct_empty ... ok
test de::tests::struct_i8 ... ok
test de::tests::struct_nothing ... ok
test de::tests::numbered_key_maps ... ok
test de::tests::struct_tuple ... ok
test de::tests::struct_option ... ok
test de::tests::struct_u8 ... ok
test de::tests::struct_with_flatten ... ok
test de::tests::tuple ... ok
test de::tests::tuple_variant ... ok
test de::unescape::tests::hex_decode_works ... ok
test de::unescape::tests::unescape_fails_for_broken_surrogates ... ok
test de::tests::wot ... ok
test de::unescape::tests::unescape_fails_for_control_characters ... ok
test de::unescape::tests::unescape_fails_for_invalid_escape_sequence ... ok
test de::unescape::tests::unescape_fails_for_lone_surrogates ... ok
test de::unescape::tests::unescape_works ... ok
test de::unescape::tests::unescape_works_for_surrogate_pairs ... ok
test ser::tests::array ... ok
test ser::tests::collect_str_can_be_used_in_custom_seralize_impl ... ok
test ser::tests::bool ... ok
test ser::tests::btree_map ... ok
test ser::tests::enum_mixed ... ok
test ser::tests::enum_variants_c_like_structs ... ok
test ser::tests::enum_variants_tuple_like_structs ... ok
test ser::tests::enum_variants_unit_like ... ok
test ser::tests::newtype ... ok
test ser::tests::invalid_json_key ... ok
test ser::tests::hash_map ... ok
test ser::tests::number ... ok
test ser::tests::str ... ok
test ser::tests::struct_ ... ok
test ser::tests::struct_bool ... ok
test ser::tests::number_key ... ok
test ser::tests::struct_i8 ... ok
test ser::tests::struct_option ... ok
test ser::tests::struct_u8 ... ok
test ser::tests::serialize_embedded_enum ... ok
test ser::tests::struct_with_flatten ... ok
test ser::tests::tuple ... ok
test test::untagged ... ok
test test::can_serde ... ok
test test::untagged_structures ... ok

test result: ok. 54 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests serde-json-wasm

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```